un_sdg/core.py

```diff
@@ -0,0 +1,184 @@
+        (all_series['Units_long'].str.contains("KG")) | (all_series['Units_long'].str.contains("Kilograms")),
+        (all_series['Units_long'].str.contains("USD")) | (all_series['Units_long'].str.contains("usd"))]   
+    
+    choices = ["%", "kg", "$"]
+    
+    all_series['short_unit'] = np.select(conditions, choices, default = None)
+    return(all_series)
+
+def get_series_with_relevant_dimensions(data_filtered,DIMENSIONS, NON_DIMENSIONS):
+    """ For a given indicator and series, return a tuple:
+    
+      - data filtered to that indicator and series
+      - names of relevant dimensions
+      - unique values for each relevant dimension
+    """
+   # data_filtered = original_df[(original_df.Indicator == indicator) & (original_df.SeriesCode == series)]
+    non_null_dimensions_columns = [col for col in DIMENSIONS if data_filtered.loc[:, col].notna().any()]
+    dimension_names = []
+    dimension_unique_values = []
+    
+    for c in non_null_dimensions_columns:
+        print(non_null_dimensions_columns)
+        uniques = data_filtered[c].unique()
+        if len(uniques) > 1: # Means that columns where the value doesn't change aren't included e.g. Nature is typically consistent across a dimension whereas Age and Sex are less likely to be. 
+            dimension_names.append(c)
+            dimension_unique_values.append(list(uniques))
+    return (data_filtered[data_filtered.columns.intersection(list(NON_DIMENSIONS)+ list(dimension_names))], dimension_names, dimension_unique_values)
+
+def generate_tables_for_indicator_and_series(data_filtered, DIMENSIONS, NON_DIMENSIONS): 
+    tables_by_combination = {}
+    dim_dict = dimensions_description()    
+    data_filtered, dimensions, dimension_values = get_series_with_relevant_dimensions(data_filtered, DIMENSIONS, NON_DIMENSIONS)
+    print("PRINTING DIMENSIONS ",dimensions)
+    print("INDICATOR", data_filtered.Indicator[0:1])
+    print("SERIES_CODE", data_filtered.SeriesCode[0:1])
+    if (len(dimensions) == 0 )|(data_filtered[dimensions].isna().sum().sum() > 0): # not the best solution.
+        # no additional dimensions
+        export = data_filtered
+        return export
+    else:
+        dim_desc = dim_dict.set_index('id').loc[dimensions].set_index('code').squeeze().to_dict()
+       #dimension_values[0] = [x for x in dimension_values[0] if x == x] # some cases where there are NaNs in the dimensions_values
+        i = 0
+        for i in range(len(dimension_values)): 
+            dimension_values[i] = [dim_desc[k] for k in dimension_values[i]]
+        for dim in dimensions:
+            print(dim)
+            data_filtered[dim] = data_filtered[dim].apply(lambda x: dim_desc[x])
+        for dimension_value_combination in itertools.product(*dimension_values):
+            # build filter by reducing, start with a constant True boolean array
+            filt = [True] * len(data_filtered)
+            for dim_idx, dim_value in enumerate(dimension_value_combination):
```

Comment:
This looks nicely terse and clever but I don't understand what exactly it is trying to do. You are eliminating certain points/columns (?) from the cross product I think but I can't follow this without more comments